fix: Add key vault to storage account - `avm/res/storage/storage-account`

Agazoth opened the pull request:
## Description

<!--
This contribution adds a possibility to output keys and connection strings to an existing key vault.
This mitigates the need for nested deployments to avoid leaking secrets to deployment logs. It also enables proper management of secrets from storage accounts.

Had to change the current test on main to exclude the Custom.Azure.Storage.UseReplication - which breaks the main pipeline.

Fixes #1780
-->

## Pipeline Reference


| Pipeline |
| -------- |
|   [![avm.res.storage.storage-account](https://github.com/Agazoth/bicep-registry-modules/actions/workflows/avm.res.storage.storage-account.yml/badge.svg?branch=KVToStorageAccount)](https://github.com/Agazoth/bicep-registry-modules/actions/workflows/avm.res.storage.storage-account.yml) |

## Type of Change

<!-- Use the check-boxes [x] on the options that are relevant. -->

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [ ] Azure Verified Module updates:
  - [x] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

<!--  Please keep up to day with the contribution guide at https://aka.ms/avm/contribute/bicep -->

eriqua commented on avm/res/document-db/database-account/gremlin-database/graph/main.json:
Please keep one PR mapped to a single module, to allow proper review and approval by the right module owner

Agazoth replied:
I reverted the faulty merged files. Should be ok now.

AlexanderSehr replied:
Please note, I'll close and re-open the PR to have it re-evaulate the rules.

AlexanderSehr replied:
Interesting, ... that did nothing. I could've sworn this worked in the past.

AlexanderSehr commented on avm/res/storage/storage-account/blob-service/main.bicep:
```diff
@@ -12,7 +12,7 @@ param automaticSnapshotPolicyEnabled bool = false
-@minValue(0)
+@minValue(1)
 @maxValue(146000)
 @description('Optional. Indicates whether change feed event logging is enabled for the Blob service. Indicates the duration of changeFeed retention in days. A "0" value indicates an infinite retention of the change feed.')
```
Please be aware, the parameter's description & new `min` value don't align.

Agazoth replied:
I agree, description should be updated to align with the documentation: https://learn.microsoft.com/en-us/azure/templates/microsoft.storage/storageaccounts/blobservices?pivots=deployment-language-bicep#changefeed

Agazoth replied:
Description updated

Agazoth commented on avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml:
```diff
@@ -81,3 +81,4 @@ rule:
     - Azure.VM.UseHybridUseBenefit
     - Azure.Storage.UseReplication
     - Azure.AppConfig.PurgeProtect
+    - Custom.Azure.Storage.UseReplication # Custom rule not working as expected
```
I have removed Custom.Azure.Storage.UseReplication from the exceptoion rules and updated the custom rule to include the correct sku name:
![image](https://github.com/Azure/bicep-registry-modules/assets/15127301/ed43f005-641a-4a98-877f-d965ebedf760)

That seems to do the trick.

eriqua commented on avm/res/storage/storage-account/README.md:
```diff
@@ -19,6 +19,7 @@ This module deploys a Storage Account.
 | `Microsoft.Authorization/locks` | [2020-05-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2020-05-01/locks) |
 | `Microsoft.Authorization/roleAssignments` | [2022-04-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2022-04-01/roleAssignments) |
 | `Microsoft.Insights/diagnosticSettings` | [2021-05-01-preview](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Insights/2021-05-01-preview/diagnosticSettings) |
+| `Microsoft.KeyVault/vaults/secrets` | [2022-07-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.KeyVault/2022-07-01/vaults/secrets) |
```
Same comment of the cognitive PR https://github.com/Azure/bicep-registry-modules/pull/1932#pullrequestreview-2056624383

Agazoth replied:
I absolutely agree with your comment on #1932 - i have startet this discussion: #1934

jikuja commented:
How does this work on following scenarios:

* re-deployment
* re-deployment after storage account key rotation?

AlexanderSehr commented:
> How does this work on following scenarios:
> 
> * re-deployment
> * re-deployment after storage account key rotation?

Presumably, all secrets would be overwritten/updated to latest each time the deployment runs

AlexanderSehr commented:
Hey @Agazoth,
for reference (I may not have mentioned it before), we documented the interface [here](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/#secrets-export). 💪

Agazoth commented:
> Hey @Agazoth, for reference (I may not have mentioned it before), we documented the interface [here](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/#secrets-export). 💪 Also updated DocumentDB (the original first module with the feature) since then.

Thanks a bunck @AlexanderSehr - I have updated the code by following your documentation - it woerked like a charm :-)

It should be ready for production now.

AlexanderSehr commented on avm/res/storage/storage-account/main.bicep:
```diff
@@ -620,6 +623,52 @@ module storageAccount_tableServices 'table-service/main.bicep' = if (!empty(tabl
+    keyVaultName: last(split(secretsExportConfiguration.?keyVaultResourceId ?? '//', '/'))
+    secretsToSet: union(
+      [],
+      contains(secretsExportConfiguration!, 'Key1Key')
```
I'd probably suggest to call it something like `accessKey1`. What was the idea behind `Key*Key`?

AlexanderSehr replied:
Similar ask for the connectionString

AlexanderSehr commented on avm/res/storage/storage-account/main.bicep:
```diff
@@ -628,6 +631,52 @@ module storageAccount_tableServices 'table-service/main.bicep' = if (!empty(tabl
+    keyVaultName: last(split(secretsExportConfiguration.?keyVaultResourceId ?? '//', '/'))
+    secretsToSet: union(
+      [],
+      contains(secretsExportConfiguration!, 'key1')
```
```suggestion
      contains(secretsExportConfiguration!, 'accessKey1')
```

AlexanderSehr replied:
Would need update in other places too

AlexanderSehr commented on avm/res/storage/storage-account/main.bicep:
```diff
@@ -628,6 +631,52 @@ module storageAccount_tableServices 'table-service/main.bicep' = if (!empty(tabl
+            }
+          ]
+        : [],
+      contains(secretsExportConfiguration!, 'key2')
```
```suggestion
      contains(secretsExportConfiguration!, 'accessKey2')
```

AlexanderSehr replied:
Would need update in other places too

AlexanderSehr commented on avm/res/storage/storage-account/tests/e2e/kvSecrets/dependencies.bicep:
```diff
@@ -0,0 +1,21 @@
+}
+
+@description('The name of the Key Vault created.')
+output keyVaultId string = keyVault.id
```
```suggestion
output keyVaultResouceId string = keyVault.id
```

AlexanderSehr commented on avm/res/storage/storage-account/main.bicep:
```diff
@@ -906,13 +906,13 @@ type secretsExportConfigurationType = {
   keyVaultResourceId: string
 
   @description('Optional. Default to Key1Key. The primary key secret name to create.')
```
Please also update each description

AlexanderSehr commented:
Nice, looks good to me. @fblix / @ktremain any doubts from your side?

ktremain commented:
@Agazoth can you please update this again from the current main and bump the version, as a recent commit now has a matching version number as your code.

Agazoth commented:
@ktremain done :-)

ktremain commented:
@Agazoth Thanks for updating the branch, please can you also increase the version number in version.json (same folder as the main.bicep), otherwise the back-end automation wont see it as a new release to publish to the Bicep module registry.

Agazoth commented:
@ktremain sorry about that 😄 - now I'm done ...

ktremain commented:
@Azure/alz-core-team - Erika added the Do not merge label previously - is it ok to remove this now? the PR looks good to me.

AlexanderSehr commented:
It's good now. The label was previously added because the schema was not defined & [documented](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/#secrets-export) yet.

On that note @ktremain, just two quick comments on what you mentioned
- 'Azure/alz-core-team' is not related to AVM
- Any PR that contains changes to module files (most notably the `main.json`) will result in a new version. However, as it would just be a new patch/bug version (<major>.<minor>.<patch>) I agree that it makes sense to update the <minor> in this case
